File: 卡码/代码随想录/1/02-链表/Go/03-反转链表/LeetCode_0206.go
//206. 反转链表
package main

import (
	"fmt"
)

type ListNode struct{
	Val int
	Next *ListNode
}

//双指针法
func reverseList1(head *ListNode) *ListNode {
	//链表为空或者只有一个节点
	if head==nil||head.Next==nil {
		return head
	}
	var pre *ListNode=nil
	cur:=head
	for cur!=nil {
		tmp:=cur.Next
		cur.Next=pre
		pre=cur
		cur=tmp
	}
	return pre
}

//递归
func reverseList(head *ListNode) *ListNode {
	//链表为空或者只有一个节点
	if head==nil||head.Next==nil {
		return head
	}
	node:=reverseList(head.Next)
	head.Next.Next=head
	head.Next=nil
	return node
}

File: 卡码/代码随想录/Python/链表/03-反转链表/LeetCode_0206.py
#206. 反转链表
from typing import *

class ListNode:
	def __init__(self,val=0,next=None):
		self.val=val
		self.next=next

class Solution:
	#递归
	def reverseList(self,head:Optional[ListNode])->Optional[ListNode]:
		#如果链表为空或者只有一个元素
		if head==None or head.next==None:
			return head
		node:Optional[ListNode]=self.reverseList(head.next)
		head.next.next=head
		head.next=None
		return node
	#双指针法
	def reverseList1(self,head:Optional[ListNode])->Optional[ListNode]:
		#如果链表为空或者只有一个元素
		if head==None or head.next==None:
			return head
		pre:Optional[ListNode]=None
		cur:Optional[ListNode]=head
		while cur:
			tmp:Optional[ListNode]=cur.next
			cur.next=pre
			pre=cur
			cur=tmp
		return pre
